optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/bi/BiConstraintStream.java

```diff
@@ -196,7 +198,42 @@
      * {@link TriJoiner joiners} are true
      */
     default <C> TriConstraintStream<A, B, C> join(Class<C> otherClass, TriJoiner<A, B, C>... joiners) {
```

Comment:
**I do not know how we missed this**, but until now, filtering joiners were not supported for tri/quad joins. (The API allowed it, but threw a runtime exception.)

This PR added test coverage that identified the problem, and so I also add the fix. The fix is to bring the code to function equivalence with bi joins, which always supported filtering joiners.

Reply:
Would they Sonar test coverage report allowed us to detect this test coverage gap? If so it might be worth checking if all api classes have any uncovered lines that matter enough to warren test coverage.

Reply:
Yeah, this would have been caught if we had that test. We do now.